### PR TITLE
[NFC] Add DIExpression::calculateFragmentIntersect

### PR DESCRIPTION
**Pull request by OCHyams**
Patch [3/x] to fix structured bindings debug info in SROA.

This function computes a fragment, bit-extract operation if needed, and new
constant offset to describe a part of a variable covered by some memory.

This generalises, simplifies, and replaces at::calculateFragmentIntersect. That
version is still used as a wrapper for now though to keep this change NFC.

The new version takes doesn't have a DbgRecord parameter, instead using an
explicit address and address offset. The old version only operates on
dbg_assigns and this change means it can also operate on dbg_declare records
easily, which it will do in a subsequent patch.

The new version has a new out-param OffsetFromLocationInBits which is set to the
difference between the first bit of the variable location and the first bit of
the memory slice. This will be used in a subsequent patch in SROA to determine
the new offset to use in the address expression after splitting an alloca.

---

Based on top of #97705 and #97719. ~Ignore the first two commits, which are under review over there.~ (rebased)

### CODE REVIEW

**Review by SLTozer — COMMENTED** [review]
`calculateFragmentIntersect` took a bit of time to properly parse, not sure I have any suggestions to improve the clarity right now (besides a minor naming suggestion); other than that this LGTM.

File: llvm/lib/IR/DebugInfoMetadata.cpp [hunk 1]
```diff
@@ -2052,6 +2052,75 @@ std::optional<DIExpression *> DIExpression::createFragmentExpression(
...
+
+  // Work towards creating SliceOfVariable which is the bits of the variable
+  // that the memory region covers.
+  // 0   4   8   12   16 ...
+  // |       |
+  // dbg location start with VarFrag offset=32
+  //         |
+  //         mem slice start: SliceOfVariable offset=40
+  int64_t MemStartRelToFragInBits =
+      MemStartRelToDbgStartInBits + VarFrag.OffsetInBits;
```
[the remark this answers is not in the document]
Comment (OCHyams) [reply llvm/lib/IR/DebugInfoMetadata.cpp:2142]: Nice catch. Change Frag to Var. Still not an amazing name, but I found coming up with meaningful names in this function was pretty difficult (hence all the comments!). Happy to take further suggestions?

**Review by OCHyams — COMMENTED** [review]
tyvm for the review - the comments have been addressed

File: llvm/lib/IR/DebugInfoMetadata.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (SLTozer) [reply llvm/lib/IR/DebugInfoMetadata.cpp:2142]: No better suggestions unfortunately, "Var" was the best I had too. I think it's reasonable, it's just _slightly_ nonsensical in that the variable isn't a region of memory that the memory slice can exist relative to, but it at least makes sense in terms of how we ultimately calculate the variable fragment covered by the slice.

**Review by SLTozer — APPROVED** [review]
Changes LGTM!

**Pull request merged**